src/java/org/apache/cassandra/index/sai/utils/BM25Utils.java

```diff
@@ -186,18 +144,15 @@ public static CloseableIterator<PrimaryKeyWithSortKey> computeScores(CloseableIt
         {
             var tf = docIterator.next();
             documents.add(tf);
-            if (docStats.avgDocLength < 0)
+            if (isOldFormat)
                 totalTermCount += tf.termCount();
         }
 
-        // avgDocLength is unknown since an index format is before {@link Version#ED},
-        // which doesn't store the total term count on disk to read it back.
-        // In such a case the old way of computing avgDocLength is used.
-        if (docStats.avgDocLength < 0)
-        {
-            assert !Version.current().onOrAfter(Version.ED) : "on Version.ED avgDocLength must be known";
-            docStats.avgDocLength = totalTermCount / documents.size();
-        }
+        // An index format before {@link Version#ED} doesn't store the total term count
+        // on the disk to read it back. Thus, if the average document length is unknown
+        // due to the old format version, it is calculated in the old way.
+        if (isOldFormat && docStats.getAvgDocLength() == 0 && totalTermCount > 0)
+            docStats.setAvgDocLength(totalTermCount / documents.size());
```

Comment:
Nit: this works, but I think this logic looks a bit fragile because we're updating a shared object that will be updated in other calls to `computeScores` by other segment indexes. I think it might be safer to assign a temporary variable and then pass that double to `scoreDoc`.

Reply:
Actually, this does leak state because on the first segment, `docStats.getAvgDocLength() == 0` will be true, but on subsequent segments, it'll be false.